### DISCUSSION TITLE
Nvim-Dap for Python

### DISCUSSION THREAD

**Discussion by msMatix** [LazyVim/LazyVim#1448] [Q&A]
Hello together,

I would like to use the debug adapter for python. For this I have made the following as in the docs:
1. adde: { import = "lazyvim.plugins.extras.dap.core" },
2. adde: { import = "lazyvim.plugins.extras.lang.python" },
in "lua/config/lazy.lua". After that I install the respective plugins. I also installed debugpy via mason.

To test the whole thing I have created an example project for myself I created a folder "test" and a "main.py" file. If I now add breakpoints, and want to start the debugger, I get no representation of the variables.

Procedure is for this:
1. "leader db" -> Toggle Breakpoint
2. "leader du" -> Dap UI
3. "leader dc" -> Debug Continue

Also starting a session with "leader ds" shows no change. Could it be that a suitable "launch.json" file is missing for this? If this is the case, are there any examples ? 

Thanks a lot already!

Matix

**Comment by dpetka2001** [comment]
I could not reproduce this with a fresh starter template.

**Reply by msMatix** [reply]
Ok. So you did the same with no Launch.json file ?

I should probably mention that I have to use Windows at work. Could that be the problem ? What operating system do you use ?

**Reply by dpetka2001** [reply]
I'm using Linux and I don't have any `launch.json` file. Just a simple `.py` file in the middle of nowhere. I also tried it in a project structure with `pyproject.toml` available and it also worked, but I don't think that should be of any importance.

**Reply by msMatix** [reply]
Or would it be possible to briefly see / show your setup ?

**Reply by dpetka2001** [reply]
I tried it both with my personal config and with a new Lazyvim starter template (by only importing `dap.core` and `lang.python`) and it worked in both cases. Can you test with a fresh Lazyvim starter template to see if it works? If it does, then it must be something in your personal configuration.

**Reply by dpetka2001** [reply]
My python setup is [here](https://github.com/dpetka2001/dotfiles/blob/main/dot_config/nvim/lua/plugins/extras/lang/python.lua), but it's mostly the same like `lang.python`. I just had started it before it even became an extra and just added some additional features that appeared in the extra (like `venv-selector`).

**Reply by msMatix** [reply]
Unfortunately, it was a fresh lazy vim starter without any customization. I would test the whole thing again tomorrow and also post the corresponding error message. 

Thanks in advance !

**Reply by dpetka2001** [reply]
Since you mentioned Windows, one problem that crossed my mind is that [nvim-dap](https://www.lazyvim.org/extras/lang/python#nvim-dap) sets the path for `debugpy`. Could in windows that path be different? Maybe uses backslash instead of slash? Maybe check that out? And also you might find some more info in the Log file, which in Linux is stored in `~/.cache/nvim/dap.log`, you might want to check what the windows location is accordingly.

**Reply by dpetka2001** [reply]
Somewhere in your personal config under `/lua/plugins/` directory you could try the following in a file
```lua
return {
  {
    "mfussenegger/nvim-dap",
    optional = true,
    dependencies = {
      "mfussenegger/nvim-dap-python",
      -- stylua: ignore
      keys = {
        { "<leader>dPt", function() require('dap-python').test_method() end, desc = "Debug Method" },
        { "<leader>dPc", function() require('dap-python').test_class() end, desc = "Debug Class" },
      },
      config = function()
        local path = require("mason-registry").get_package("debugpy"):get_install_path()
        if vim.loop.os_uname().sysname:find("Windows") then
          require("dap-python").setup(path .. "\\venv\\bin\\python")
        else
          require("dap-python").setup(path .. "/venv/bin/python")
        end
      end,
    },
  },
}
```
This takes into consideration the Windows path as well, in case that might be the problem. Test it out to see if it works.

**Comment by msMatix** [comment]
Hey,

I have checked both once and unfortunately this does not work either.
When I start the debug session I get the following warning: 

**Debug adapter didn't respond. either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or 'python' configuration.**

After that a cmd-line opens with the following content:

> Content-Length: 137
> 
> {"seq": 1, "type": "event", "event": "output", "body": {"category": "telemetry", "output": "ptvsd", "data": {"packageVersion": "1.8.0"}}}Content-Length: 139
> 
> {"seq": 2, "type": "event", "event": "output", "body": {"category": "telemetry", "output": "debugpy", "data": {"packageVersion": "1.8.0"}}}

Unfortunately there is nothing useful in the Dap log either:

> [ ERROR ] 2023-09-14T08:16:38Z+0200 ] ...ppData/Local/nvim-data/lazy/nvim-dap/lua/dap/session.lua:1426 ]	"stderr"	{
>   command = "C:\\Users\\xxx\\AppData\\Local\\nvim-data\\mason\\bin\\debugpy-adapter.CMD",
>   type = "executable"
> }	"^C"

**Reply by dpetka2001** [reply]
Well there is an error in your `dap.log` and it seems like the problem lies with `nvim-dap` or `nvim-dap-python` itself. Unfortunately, I don't use Windows to test, so your best bet would be to make an issue on either repo to get more proper support. Just to make sure, did you try my previous snippet of code in your personal configuration? Also, there seems to be some other stuff after the 4 lines with the `ERROR` that you posted. Could you also post the rest?

**Reply by msMatix** [reply]
Hey
ah that could be of course. Thanks for the hint. Yes exactly I have used your snippet and unfortunately that has also brought nothing.  With the other lines it is the same error message only at a later time. Therefore I had not inserted these more with. 

Thanks a lot for your support! :)

**Reply by dpetka2001** [reply]
Do you know the full path of where `Mason` installs packages on windows? You could try `require("dap-python").setup("/mason/full/path/to/venv/python/")` as a last resort. If that doesn't work either and the `ERROR` in your `dap.log` remains the same, you would be better off creating an issue at the repo of the adapter, as I cannot offer further help I'm afraid.

**Reply by msMatix** [reply]
yes I had also tried this morning and unfortunately received the same error message. Therefore I would rather ask the adapter. Thanks for your support!

**Reply by wimmstr** [reply]
I had the same Problem on Windows.
The fix is simple, however I dont know who introduced it:
this line in the nvim-dap config:
 require("dap-python").setup(path .. "/venv/bin/python")

is wrong.

"bin" needs to be "Scripts" instead.

**Reply by msMatix** [reply]
Hey,
thanks for the idea. But I find debugpy in the following folder:
`C:\Users\xxx\AppData\Local\nvim-data\mason\packages\debugpy\venv\Scripts`
Or is this the wrong path?

**Reply by dpetka2001** [reply]
Just try it out then and see if it works. I don't use Windows, so I can't check it.

**Reply by wimmstr** [reply]
Thats same Path I have on Windows. I needed to correct 
` config = function()
        local path = require("mason-registry").get_package("debugpy"):get_install_path()
          require("dap-python").setup(path .. "/venv/bin/python")
           end,`

the Path here. (/bin/ > /Scripts/)
It works then. 
.. at least until LazyVim updates, because I edit it right in the /extras/python.lua file.

What I dont know is who produces that incorrect Path? Is that handwritten by folke? Produced by nvim-dap?
Do other Windows Installs also use a "Scripts" folder instead of "bin"?
I use poetry and pyenv-win-virtualenv also. I dont know if that changes anything.

**Reply by dpetka2001** [reply]
Nobody introduced that error. It's just that the plugin `nvim-dap-python` was probably programmed under a Unix environment, so the author didn't take into account the Windows environment. The extra here in LazyVim was also made by people who use Linux, so that path got taken into account. If you want to also take into account the Windows path, then you would have to do something like what I described in my other [comment](https://github.com/LazyVim/LazyVim/discussions/1448#discussioncomment-6994889), so that both environments are taken into account. I don't use Windows, so I can't test it myself, but if you can make it work on Windows, please feel free to submit a PR that fixes the current situation. Also you shouldn't edit it in the `lazyvim/plugins/extras/python.lua` file. You should instead create a file in your `/lua/plugins/` directory and just overwrite the default `config` function. That way you won't have problems with LazyVim updates (except if something changes drastically in the default `config` function for `dap-python` spec). But if you submit a PR that fixes it, I'm sure there's a high chance it will get integrated into LazyVim.

**Reply by wimmstr** [reply]
Thanks for the detailed answer!
Ill look into nvim-dap-python source to see if it refrences the wrong path for windows and try to implement your OS check and my path fix. 
I worded it poorly, but I was trying to explain that I dont know where the PR should go, I guess =) I´m really new to programming, so im mostly winging it.
Thanks for your help!

**Reply by dpetka2001** [reply]
`nvim-dap-python` had a [commit](https://github.com/mfussenegger/nvim-dap-python/commit/e0be843877e7ae756ef1ee7a441ca0b9e1677da9) today that uses `vim.fn.has("win32")` instead of `vim.loop.os_uname().sysname:find("Windows", 1, true)`, because the `sysname` doesn't always return `Windows` according to the plugin's maintainer, so you might want to use that for your case from now on.

**Reply by rdataforge** [reply]
same issue here in windows. Followed the above steps to configure nvim-dap but not working either.
any way to solve this?

**Comment by drowning-in-codes** [comment]
I have the same issue. I install similar packages on Mac and it works fine. But on windows, when I  try to debug python, it shows 
`Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or `python` configuration. Check the logs for errors (:help dap.set_log_level)`
and it just keeps suspension and no response.  
No content in dap.log until I terminate debugging. My configurations are just purely lazyvim and required extras are downloaded.  python version 3.12.
